### ISSUE TITLE
gaussian neighborhood formula

### ISSUE THREAD

**Issue by yoch** [peterwittek/somoclu#61]
Hi,

What is the source of [this formula](https://github.com/peterwittek/somoclu/blob/master/src/mapDistanceFunctions.cpp#L70-L73) for the gaussian neighborhood : `exp ( dist^2 / ((2 (radius+1)^2) / stddevs^2) )` ? (namely what `stddevs` means ? I guess the gaussian std = radius/stddevs, but is that correct ?)

And why stddevs [is fixed to 2](https://github.com/peterwittek/somoclu/blob/master/src/mapDistanceFunctions.cpp#L80) ? For some problems a lower std works much better (by instance with std = radius * 0.3).

**Note** : the current code can be simplified as

```cpp
float gaussianNeighborhood(float distance, float radius, float stddevs) {
    float std = (radius + 1) / stddevs;
    return exp((- distance * distance) / (2 * std * std));
}
```

**Comment by peterwittek** [comment]
This is another good question. Historically, the neighbourhood function was more or less unchanged since Somoclu's predecessor, [mr-mpi-som-gpu](https://github.com/peterwittek/mr-mpi-som-gpu/blob/master/mrsom.cpp#L290), which in turn traces its roots to [mr-mpi-som](https://github.com/andreyto/mr-mpi-som/blob/master/src/mrsom.cpp#L460). How the slight modifications entered the function, I don't remember, and that hard-coded 2 is indeed fishy. The truth is I have been happy with the maps we got, and never bothered to look into it deeper.

If it makes a difference to you, introducing `std` as a parameter takes epsilon effort. I could do it over the coming days.

**Comment by yoch** [comment]
Kohonen (in SOM_PAK and SOM toolbox) uses the radius directly as std, but I've often observed that smaller values give better results. In fact, you currently use `std = (radius + 1) * 0.5` (why the +1 ?).

I think it's a good idea to introduce it as a coefficient, something like `std_coef`. The code might look like this : 

```cpp
float gaussianNeighborhood(float distance, float radius, float coef) {
    float std = (radius + 1) * coef;
    return exp((- distance * distance) / (2 * std * std));
}

float getWeight(float distance, float radius, float scaling, float coef=0.5, ...) {
    float result = 0.0;
    if (gaussian) {
        if (!compact_support || distance <= radius) {
            result = gaussianNeighborhood(distance, radius, coef);
        }
    } 
    /*...*/
    return scaling * result;
}
```

**Comment by peterwittek** [comment]
Commit f4a10d44d7f8f2ce240dd4ee58c7d7ed7c382398 exposes this parameter across all interfaces. I also made compact support default, which is the canonical way the Gaussian neighborhood functions.

**Comment by yoch** [comment]
Very helpful, thanks you very much ! I will be able to play with now :).

IMHO, in the python code `std_coef` should be a parameter for the `train` function, and not necessarly to be related to the SOM itself.

**Comment by peterwittek** [comment]
All other neighborhood-related parameters are also passed in the constructor. I feel that they belong to the model, and not to the nitty-gritty of the training (epochs, radius, learning rate, cooling method). Since we are talking about Python, you can change the private variables of an object anyway if you want to try different `std_coeff` values on the same data.

**Comment by yoch** [comment]
Very good.
